src/server/zset_family.cc

```diff
@@ -2166,64 +2181,106 @@ void ZSetFamily::ZCount(CmdArgList args, const CommandContext& cmd_cntx) {
   }
 }
 
-void ZSetFamily::ZDiff(CmdArgList args, const CommandContext& cmd_cntx) {
-  vector<vector<ScoredMap>> maps(shard_set->size());
+void ZDiffInternal(CmdArgList args, const CommandContext& cmd_cntx, bool store) {
+  vector<OpResult<vector<ScoredMap>>> maps(shard_set->size(), OpStatus::SKIPPED);
+  auto* rb = static_cast<RedisReplyBuilder*>(cmd_cntx.rb);
+  const string_view dest_key = ArgS(args, 0);
+  const ShardId dest_shard = Shard(dest_key, maps.size());
+  // Get base key from args for both commands
+  const string_view key = store ? ArgS(args, 2) : ArgS(args, 1);
+  const ShardId sid = Shard(key, maps.size());
+
   auto cb = [&](Transaction* t, EngineShard* shard) {
-    maps[shard->shard_id()] = OpFetch(shard, t);
+    // We should skip destkey with ZDIFFSTORE if shard id matches
+    const bool skip_dest_key = store ? shard->shard_id() == dest_shard : false;
+    maps[shard->shard_id()] = OpFetch(shard, t, skip_dest_key);
     return OpStatus::OK;
   };
 
-  cmd_cntx.tx->ScheduleSingleHop(std::move(cb));
+  cmd_cntx.tx->Execute(std::move(cb), !store);
 
-  const string_view key = ArgS(args, 1);
-  const ShardId sid = Shard(key, maps.size());
-  auto* rb = static_cast<RedisReplyBuilder*>(cmd_cntx.rb);
-  // Extract the ScoredMap of the first key
-  auto& sm = maps[sid];
-  if (sm.empty()) {
-    rb->SendEmptyArray();
-    return;
+  // Check if any shard contains WRONG_TYPE error and report error
+  for (auto& shards : maps) {
+    if (shards.status() == OpStatus::WRONG_TYPE) {
+      if (store) {
+        cmd_cntx.tx->Conclude();
+      }
+      return rb->SendError(maps[sid].status());
+    }
   }
-  auto result = std::move(sm[0]);
-  sm.erase(sm.begin());
 
-  auto filter = [&result](const auto& key) mutable {
-    auto it = result.find(key);
-    if (it != result.end()) {
-      result.erase(it);
-    }
-  };
+  // Extract the ScoredMap of the first key
+  auto& sm = maps[sid].value();
+  vector<ScoredMemberView> smvec;
+  // If key is not empty we can iterate over all fetched keys
+  const bool key_empty = sm.empty() ? true : false;
+  if (!key_empty) {
+    auto result = std::move(sm[0]);
+    sm.erase(sm.begin());
+
+    auto filter = [&result](const auto& key) mutable {
+      auto it = result.find(key);
+      if (it != result.end()) {
+        result.erase(it);
+      }
+    };
 
-  // Total O(L)
-  // Iterate over the results of each shard
-  for (auto& vsm : maps) {
-    // Iterate over each fetched set
-    for (auto& sm : vsm) {
-      // Iterate over each key in the fetched set and filter
-      for (auto& [key, value] : sm) {
-        filter(key);
+    // Total O(L)
+    // Iterate over the results of each shard
+    for (auto& vsm : maps) {
+      // Skip shard if empty
+      if (vsm.status() == OpStatus::SKIPPED)
```

Comment:
It could be that the keys do not exist (find in opfetch will return empty iterator) and therefor the vector here will be empty but status is not skipped. What is the value of using skipped status? why do we need it? why not just return empty vector ?

Reply:
We initialize `maps` vector to number of shards in system, initially status for each shard is set to `SKIPPED`. We will skip shard that are is not used in diff computation (either by not using them or if destination key is only found).

Reply:
Note that we will skip the shards that do not have keys not because you set the skipped but because the callback will not be executed on this shard. The transaction framework will make sure that the callback is executed only on the relevant shards.
I think that the skipped redundant as the map vector will be empty and this add minor complexity to the code. I dont think its very important comment but more of my approach